### ISSUE TITLE
possible bug when appending to string

### ISSUE THREAD

**Issue by krux02** [nim-lang/Nim#5913]
```Nim
var str = "abcdef"
echo cast[ptr tuple[length,capacity: int]](str)[]
str.add 'g'
echo cast[ptr tuple[length,capacity: int]](str)[]
str.add 'h'
echo cast[ptr tuple[length,capacity: int]](str)[]
```

output:
```
(length: 6, capacity: 7)
(length: 7, capacity: 7)
(length: 8, capacity: 14)
```

almost the same in c++

```C++
#include <iostream>
#include <string>

using namespace std;

int main() {
  string str = "somestufftocircumventsmallstringoptimizations -- abcdef";
  cout << "length: " << str.length() << " capacity: " << str.capacity() << endl;
  str.push_back('g');
  cout << "length: " << str.length() << " capacity: " << str.capacity() << endl;
  str.push_back('h');
  cout << "length: " << str.length() << " capacity: " << str.capacity() << endl;
  str.push_back('i');
  return 0;
}
```

output:
```
length: 55 capacity: 55
length: 56 capacity: 110
length: 57 capacity: 110
```
As you can see, in the c++ version (using g++) the doubling of the string capacity happens directly after I insert the first character to the string. In Nim this does not happen directly. It happens after I insert he second character. I am not 100% sure about the Nim implementation of the string type, but this is a big hint for me, that there is a bug with the Nim implementation where the terminating ``'\0'`` is not treated properly.

**Comment by Araq** [comment]
So ... what IS the bug? The invariant is ``len <= cap`` and both do not count the \0 iirc.

**Comment by krux02** [comment]
When both do not count the ``'\0'`` as you said, then the bug is that string literals always use one byte too much memory. Ok, that is technically not a bug, but something that should be fixed in my opinion.

**Comment by Araq** [comment]
How do they use one byte too much? The terminating zero is required.

**Comment by krux02** [comment]
This is the length and capacity of a literal wtih the terminating ``'\0'``, I did not do anything on the string:
`(length: 6, capacity: 7)`

Here I appended one character that that exact same string: 
`(length: 7, capacity: 7)`
The capacity did not double in size. So either the terminating ``'\0'`` got overriden and it is a bug. Or it is harmless and the initial capacity was allocated with one byte more than necessary.
